File: codility/max_prod3.cc
// 100%
#include <algorithm>
int solution(vector<int> &A) {
    int len = A.size();
    
    sort(A.begin(), A.end());
    
    // covers cases with negative numbers and 0s
    // the product of 1st 2 neg. numbers will be positive
    return max(A.front() * A[1] * A.back(), A[len-3] * A[len-2] * A.back());
}

// 0% (passed example only)
// fails:
// [-1, -1, 1, 7] returns -7 instead of 7
// [-1, -1, 0, 1, 7] returns 0 instead of 7
#include <algorithm>
int solution(vector<int> &A) {
    // write your code in C++98
    long int prq;
    int n = 3;
    sort(A.begin(), A.end());
    
    prq = A.back();
    for (int i = A.size()-2; i > 0 && n > 0; i--) {
        prq *= A[i];
        --n;
    }
    return prq;
}

File: archive/codility/lesson_4_easy/max_prod3.py
# 100% both
def solution(A):
    n = len(A)
    A.sort()
    # if many negative numbers, multiplying neg. gives positive
    return max(A[0] * A[1] * A[n-1], A[n-3] * A[n-2] * A[n-1])
    # or using negative indices
    #return max(A0] * A[1] * A[-1], A[-3] * A[-2] * A[-1])